BUG: Fix overflow bugs in date_Range

jbrockmendel opened the pull request:
- [x] closes #24123
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Fixing this turned out to be a small beast.  The overflows of interest fall into three categories.  First: overflows when computing `periods * stride` that ATM are _not_ caught, so give incorrect results:

```
>>> pd.date_range('1969-05-04', periods=200000000, freq='30000D')
DatetimeIndex(['1969-05-04', '2051-06-23', '2133-08-12'], dtype='datetime64[ns]', freq='30000D')

>>> dti = pd.date_range(start='1677-09-22', end='2262-04-11', freq='D')
>>> pd.date_range(start=dti[0], periods=len(dti), freq='D')
DatetimeIndex([], dtype='datetime64[ns]', freq='D')
```

Second, overflows when `periods * stride` overflow the int64 bounds but not uint64 bounds, which in some cases are recoverable.  The following should _not_ raise:

```
>>> dti = pd.date_range(start='1677-09-22', end='2262-04-11', freq='D')
>>> pd.date_range(start=dti[0], periods=len(dti)-1, freq='D')
[...]
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Cannot generate range with start=-9223286400000000000 and periods=213502
```

Last is a similar one with a negative stride where we can recover, but not by converting to uint64:

```
>>> dti = pd.date_range(start='1677-09-22', end='2262-04-11', freq='D')
>>> pd.date_range(end=dti[-1], periods=len(dti), freq='D')
[...]
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Cannot generate range with end=9223372800000000000 and periods=213503
```

jbrockmendel committed:
Fix overflow bugs in date_Range

jbrockmendel committed:
add GH reference

mroeschke commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1747,12 +1748,78 @@ def _generate_range_overflow_safe(endpoint, periods, stride, side='start'):
     assert side in ['start', 'end']
+
+    i64max = np.iinfo(np.int64).max
+    msg = ('Cannot generate range with {side}={endpoint} and '
```
Mentioned in https://github.com/pandas-dev/pandas/issues/24252#issuecomment-446729164, would it be a good opportunity to change this error message so we display is date being overflowed instead of the nanosecond epoch time (since that was mostly likely the original argument)?

jbrockmendel replied:
Good idea

mroeschke commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1762,6 +1829,33 @@ def _generate_range_overflow_safe(endpoint, periods, stride, side='start'):
+    mid_periods = periods // 2
+    remaining = periods - mid_periods
+    assert 0 < remaining < periods, (remaining, periods, endpoint, stride)
+    print(periods, mid_periods, endpoint, stride, side)
```
Leftover `print` statement?

jbrockmendel replied:
Yep, will fix

jreback reviewed (changes requested):
- why are you not using ``algos.checked_add_with_arr``?
- I am -1 on using uint64 here at all. why add all of this complexity, simply raise?

jbrockmendel commented:
> why are you not using algos.checked_add_with_arr?

We are in one place, but it is mostly unnecessary, adds a lot of machinery for NA-masking that isn't relevant here.

> I am -1 on using uint64 here at all. why add all of this complexity, simply raise?

Why would we want to raise on valid inputs?  It's entirely possible that there is a simpler way to do this; I banged my head against it for a while before I got it working, so by the end elegance was not a priority.

These overflows were found when debugging the pickle problems in #24074.  There is a test that creates a DTI (something like) `dti = pd.date_range(start='1677-09-22', end='2262-04-11', freq='D')`, pickles it, then tries to unpickle it.  In the unpickling, it calls validate_frequency, which calls `cls._generate_range(start=dti[0], periods=len(dti), freq=dti.freq)`, which in master currently returns an incorrect result because of a silent overflow in multiplication.

jbrockmendel committed:
cleanup, simplification, better error message

jreback commented:
> Why would we want to raise on valid inputs? It's entirely possible that there is a simpler way to do this; I banged my head against it for a while before I got it working, so by the end elegance was not a priority

how are these valid inputs iif you are overflowing int64? I would rather not jump thru hoops and have over complex code.

jbrockmendel commented:
> how are these valid inputs iif you are overflowing int64?

Because the biggest number we can get by adding an int64 to `np.iinfo(np.int64).min` is -1, but we can have date_ranges that start near the lower implementation bound and proceed well past 1970-01-01.

I claim that for _any_ DTI produced with `dti = pd.date_range(start=start, end=end, freq=freq)`, we should be able to write `dti2 = pd.date_range(start=dti[0], periods=len(dti), freq=dti.freq)` and have `assert dti2.equals(dti)`.  Ditto for `dti3 = pd.date_range('end=dti[-1], periods=len(dti), freq=dti.freq)`.

jbrockmendel committed:
handle missed corner case

jbrockmendel commented:
Travis fail is Hypothesis

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into over

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1747,7 +1747,8 @@ def _generate_regular_range(cls, start, end, periods, freq):
+                                  side='start'):
     """
     Calculate the second endpoint for passing to np.arange, checking
     to avoid an integer overflow.  Catch OverflowError and re-raise
```
can you doc-string this a bit more (say what the parameters mean)

jbrockmendel replied:
sure

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1785,6 +1852,32 @@ def _generate_range_overflow_safe(endpoint, periods, stride, side='start'):
+    periods : int
+    stride : int
+    side : {'start', 'end'}
+
```
so pls move all of this new code to the same place as checked_add_with_arr

jbrockmendel replied:
algos seems like a weird place to put datetime-specific code, no?  I guess in principle this code could be used for generating int64 ranges...

jreback commented on pandas/tests/indexes/datetimes/test_date_range.py:
```diff
@@ -88,6 +88,33 @@ def test_date_range_nat(self):
         with pytest.raises(ValueError, match=msg):
             date_range(start=pd.NaT, end='2016-01-01', freq='D')
 
+    def test_date_range_multiplication_overflow(self):
```
so 2 test cases *fully* exercise all of this code???

jbrockmendel replied:
Good call, I'm tracking down untested cases now.

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1770,12 +1771,78 @@ def _generate_range_overflow_safe(endpoint, periods, stride, side='start'):
+                if result <= i64max + np.uint64(stride):
+                    return result
+            else:
+                return _generate_range_recurse(endpoint, periods,
```
what are these 2 cases?

why is this recursion? why can't you do this iteratively?

jbrockmendel replied:
> why is this recursion? why can't you do this iteratively?

`generate_range_recurse` breaks the problem down into two smaller pieces and calls `_generate_range_overflow_safe` twice.  It is iterative in the sense that it makes two calls, as opposed to repeated calls to itself.  It is recursive inasmuch as it is called via `_generate_range_overflow_safe`, which it also calls.  I'm open to other ideas for names.

> what are these 2 cases?

Which two cases are you asking about?

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into over

jbrockmendel committed:
more tests, docstrings

jbrockmendel commented:
Updated with a couple of tests for previously un-hit cases, removed some code that turned out to be un-reachable, fleshed out docstring params.

It turns out the call to `np.arange` can overflow.  Now catch that too (with test)

jbrockmendel commented:
gentle ping.  I think this will fix the inability to do frequency validation discussed in #24024.

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1756,9 +1764,13 @@ def _generate_range_overflow_safe(endpoint, periods, stride, side='start'):
     Parameters
     ----------
     endpoint : int
```
I would like you to move this out of this file; you can make a subdirectory ``pandas/core/arrays/datetimes`` if you want or make this a private module

jbrockmendel replied:
OK.  Maybe `arrays._ranges`?  Also opens a natural place if we ever want to have a RangeArray.

could also move timedeltas._generate_regular_range there (and looking at it, it seems plausible it would have the same bug as the datetime version.

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1728,7 +1728,14 @@ def _generate_regular_range(cls, start, end, periods, freq):
 
-        values = np.arange(b, e, stride, dtype=np.int64)
+        with np.errstate(over="raise"):
+            try:
```
also need to be moved out to another module, this is adding way to much reader complexity here (reader means future contributor to pandas)

jbrockmendel replied:
OK.  I think this can also be cleaned up by making a decorator to set np.errstate once instead of repeatedly.

jbrockmendel replied:
any thoughts on moving offsets.generate_range to this new module?  that is the only place where it is used; thematically a much better fit here than in offsets (much of which should go up to tslibs anyways)

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into over

jbrockmendel committed:
move range-generation functions to arrays._ranges

jbrockmendel committed:
avoid over-writing tz

jreback commented on pandas/core/arrays/_ranges.py:
```diff
@@ -0,0 +1,186 @@
+                             "if a 'period' is given.")
+
+        with np.errstate(over="raise"):
+            try:
```
comment on why you need this

jreback commented on pandas/core/arrays/_ranges.py:
```diff
@@ -0,0 +1,186 @@
+        return _generate_range_overflow_safe_signed(
+            endpoint, periods, stride, side)
+
+    elif ((endpoint > 0 and side == 'start' and stride > 0) or
```
isn't this case already hendled in _generate_range_overflow_safe_signed? l169

jbrockmendel replied:
No.  The check on 169 is specific to the case where periods * stride doesn't overflow int64 bounds.  The check here on 127 is for the case where it does.

That said, this check could be removed.  It is a fast-path to fail early.

jreback replied:
k make sure that all of these paths are hit and tested - make as simple as possible

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into over

jbrockmendel committed:
overflow comment

jreback commented:
thanks
